compiler/dyno/util/linters/fieldsUsed.cpp

```diff
@@ -0,0 +1,258 @@
+                                    .bind("member")))
+              .bind("method")));
+}
+
+// Run MemberExprsCallback over functions which match `pattern`.
+// Require uses of all fields of `this` that meet the `requireField`
+// filter function
+// F: const FieldDecl* -> bool
+template <typename F>
+static size_t runMemberExprsMatcher(clang::tooling::ToolExecutor* ex,
+                                    llvm::StringRef pattern, F requireField) {
+  ast_matchers::MatchFinder finder;
+  PtrPrintInfoMap infoMap;
+  MethodFieldUseMap useMap;
+  auto callback = MemberExprsCallback<F>(infoMap, useMap, requireField);
+
+  // This isn't the best query, since if we pass multiple TUs on the
+  // command line, I think we'll revisit functions, but having the
+  // query be on the leaves is a) convenient and b) driving a nested
+  // matcher is still something I'm not sure how to do
+  // TODO make the matchesName an argument
+  finder.addMatcher(makeMemberExprsMatcher(pattern), &callback);
+
+  auto err = ex->execute(newFrontendActionFactory(&finder));
+  if (err) {
+    llvm::errs() << llvm::toString(std::move(err)) << "\n";
+    return 1;
+  }
+
+  // At this point, everything in useMap is a freed pointer, but we just look up
+  // the info in the map we've returned. This is my workaround to not finishing
+  // understanding the lifetime of AST in clang in one sitting
+
+  size_t errCount = 0;
+
+  for (auto it : useMap) {
+    const PrintInfo& methodInfo = infoMap.at(it.first);
+
+    for (const void* fieldNotUsed : it.second) {
+      const PrintInfo& fieldInfo = infoMap.at(fieldNotUsed);
+      errCount += 1;
+      llvm::outs() << methodInfo.location << "\n  " << methodInfo.name
+                   << " does not use " << fieldInfo.name << "\n";
+    }
+  }
+
+  return errCount;
+}
+
+static bool allFields(const FieldDecl* fd) {
+  (void)fd;
+  return true;
+}
+
+static bool fieldsForMarkFunction(const FieldDecl* fd) {
+  auto t = fd->getType();
+  if (t->isBuiltinType() || t->isEnumeralType())
+    return false;
+  std::string name = fd->getType().getAsString();
+  auto re = llvm::Regex("std::string");
+  if (re.match(name)) {
+    return false;
+  }
+  return true;
+}
+
+// Boilerplate: Set up the command line options
+static cl::extrahelp commonHelp(CommonOptionsParser::HelpMessage);
+static cl::OptionCategory toolTemplateCategory("tool-template options");
+
+int main(int argc, const char** argv) {
+  llvm::sys::PrintStackTraceOnErrorSignal(argv[0]);
+
+  auto executor = clang::tooling::createExecutorFromCommandLineArgs(
+      argc, argv, toolTemplateCategory);
+
+  if (!executor) {
+    llvm::errs() << llvm::toString(executor.takeError()) << "\n";
+    return 1;
+  }
+
+  // This is a StandaloneToolExecutor under the hood
+  auto ex = executor->get();
+
+  size_t errCount = 0;
+
+  errCount +=
+      runMemberExprsMatcher(ex, "update$|swap$|hash$|operator==", allFields);
```

Comment:
Can it handle contentsMatchInner / *ContentsMatchInner as well as markUniqueStringsInner / *MarkUniqueStringsInner ?

Reply:
don't know, was leaving that for future work

Reply:
Future work is fine but it looks like it would be really easy to try adding it (which is a good thing)